# Exploring eBay Car Sales Data

In this project we are going to explore [used cars data](https://data.world/data-society/used-cars-data). Dataquest has made a few modifications to make the data dirtier. The aim of this project is to clean the data and analyze the used car listings.

In [116]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

Let's explore the dataset.

In [117]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [118]:
autos.head(3)

dateCrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offerType   price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000  control         bus                2004    manuell   
1   Angebot  $8,500  control   limousine                1997  automatik   
2   Angebot  $8,990     test   limousine                2009    manuell   

   powerPS   model   odometer  monthOfRegistration fuelType       brand  \
0      158  andere  150,000km                    3      lpg     peugeot   
1      286     7er  150,000km                    6   benzin         bmw   
2      102    golf   70,000km                    7   benzin  volkswagen   

  notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

## Fixing the column names

With `autos.info()` we can see useful information like column names and types. We can see that the dataset contains 20 columns. Most of the columns are string, some are integer. There are some null values, especially on the `notRepairedDamage` column. The column names use camel case instead of Python's preferred snake case, so let's begin by changing the column names to snake case.

In [119]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [120]:
# Copy the above and change the values.

autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [121]:
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   
2    Angebot  $8,990     test    limousine               2009    manuell   

   power_ps   model  kilometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

The columns are now in snakecase, which is Python's preferred naming convention. Now lets begin to explore our data

## Cleaning data

In [122]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

        kilometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

We can see that the `seller` and `offer_type` columns are mainly just one value, meaning that we can delete those columns. We can also see that some numeric data needs to be cleaned, for example `price` and `kilometer` data contains some symbols instead of numbers. `nr_of_pictures` seems to only contain 0's as values, so let's check that out to make sure.

In [123]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

As we can see there are only 0's in `nr_of_pictures`. Let's continue with deleting columns `seller`, `offer_type` and `nr_of_pictures`.

In [124]:
autos = autos.drop(["seller", "offer_type", "nr_of_pictures"], axis = 1)

Now we need to clean the columns that contained symbols instead of numeric values.

In [125]:
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(int)
autos["kilometer"] = autos["kilometer"].str.replace(",", "").str.replace("km", "").astype(int)

Let's also change the `kilometer` column's name to `odometer_km`:

In [126]:
autos = autos.rename({"kilometer": "odometer_km"}, axis=1)

We will now dive deeper into `price` and `odometer_km` columns. We are looking for data that doesn't look right, for example unrealistic prices. We are going to look for unique values, min/max/median/mean values etc.

In [127]:
autos["odometer_km"].unique().shape

(13,)

From the output we can see that there are only 13 unique values for `odometer_km`.

In [128]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

From above we can see useful information about the `odometer_km` column. There are 50,000 values, but we can also see that over 50% of the values have a value of 150,000.

In [129]:
autos["odometer_km"].value_counts().head(10).sort_index()

5000        967
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

This list shows us amount of values for each `odometer_km` value, for example there are 32424 values for 150000. This tells us that the `odometer_km` values are rounded. Let's now take a look at prices using the same methods.

In [130]:
autos["price"].unique().shape

(2357,)

This time there are a lot more unique values, 2357.

In [131]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

There are 50,000 values as expected. The minimum price is 0, which might mean that there are some rows missing a value.

In [132]:
autos["price"].value_counts().sort_index()

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

From above we can see that there are 1421 cars that were sold "for free". Prices 11111111, 12345678, 99999999 also seem unrealistic. There also seems to be a lot of cars sold for basically free, but the starting bid could be just 1, so we are going to keep those. Let's sort from the highest price to smallest.

In [133]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

Looks like prices lower than 389000 seem realistic, so we are going to use prices that are lower than that. Let's take a look at price range of 500 and 389000.

In [134]:
print(autos["price"].between(1,389000).value_counts())

True     48565
False     1435
Name: price, dtype: int64


Above we can see that there are 48565 cars sold between the price range of 1 to 389000. There are 1435 outliers that will be deleted.

In [135]:
autos = autos.loc[(autos["price"] >= 1) & (autos["price"] <= 389000)]

## Dates

There are 5 different dates in our dataset: `date_crawled`, `last_seen`, `ad_created`, `registration_month` and `registration_year`. Let's print them to see more information.

In [136]:
autos[['date_crawled','last_seen','ad_created','registration_month','registration_year']][0:3]

date_crawled            last_seen           ad_created  \
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00   
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00   
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009

As we can see the format is year-month-day. The `registration_month` and `registration_year` only contains an integer value, while the other columns are strings. We are now going to calculate the distribution of values for `date_crawled`, `ad_created` and `last_seen` columns. We can get the date by taking first 10 characters of the column.

In [137]:
date_crawled_dates = autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()
ad_created_dates = autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()
last_seen_dates = autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()
print(date_crawled_dates)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


In [138]:
print(ad_created_dates)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64


In [139]:
print(last_seen_dates)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


The dates seem to be fairly equally distributed. Let's now move on to the `registration_year`.

## Registration year

In [140]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The maximum year for registrations is 9999 and the minimum year is 1000, which means that there are incorrect values in the dataset. Also the `last_seen` column shows that the data is from before 2016, so any cars with a registration year after 2016 are incorrect. Because of this, we are going to use 2016 as our latest acceptable year. Quick Google search tells us that cars were invented in 1886, so the lowest year we are going to use is 1900.

In [141]:
(~autos["registration_year"].between(1900,2016)).sum()

1884

We can see that there are 1884 cars that are not in our acceptable set. Let's remove them. After that we will calculate the distribution for rest of the registration years.

In [142]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [143]:
registration_year_distribution = autos["registration_year"].value_counts(normalize=True).sort_index()
registration_year_distribution

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: registration_year, Length: 78, dtype: float64

This shows us that there are very few cars registrated in the lower bounds. Let's take a closer look at cars for the last 20 years of data.

In [144]:
registration_year_distribution.tail(20)

1997    0.041794
1998    0.050620
1999    0.062060
2000    0.067608
2001    0.056468
2002    0.053255
2003    0.057818
2004    0.057904
2005    0.062895
2006    0.057197
2007    0.048778
2008    0.047450
2009    0.044665
2010    0.034040
2011    0.034768
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: registration_year, dtype: float64

From 1997 to 2016 the amounts are fairly evenly distributed, newer cars have a little lower distribution.

## Brands

Let's now take a look at different car brands. Let's start by making a distribution for each car brand.

In [145]:
brand_distribution = autos["brand"].value_counts()
print(brand_distribution.describe())
brand_distribution

count      40.000000
mean     1167.025000
std      1993.654826
min        27.000000
25%       114.250000
50%       396.500000
75%       787.750000
max      9862.000000
Name: brand, dtype: float64


volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

There are 40 different car brands. Above we can see them all sorted from highest value to lowest value. From now on we are only going to focus on the top 10, which are:

In [146]:
brand_distribution[:10]

volkswagen       9862
bmw              5137
opel             5022
mercedes_benz    4503
audi             4041
ford             3263
renault          2201
peugeot          1393
fiat             1197
seat              853
Name: brand, dtype: int64

Now we are going to take a look at the mean price for each of the top 10 brands.

In [147]:
top_10_brands = brand_distribution[:10].index
top_10_price_mean = {}

for b in top_10_brands:
    top_10_price_mean[b] = autos.loc[autos["brand"] == b, "price"].mean()

top_10_price_mean

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287,
 'renault': 2474.8646069968195,
 'peugeot': 3094.0172290021537,
 'fiat': 2813.748538011696,
 'seat': 4397.230949589683}

From the dictionary above we can see that from top 10 brands the most expensive ones are Audi, Mercedes-Benz and BMW. Top 3 cheapest are Renault, Fiat and Opel.

## Average mileage

Next we are going to look if cars average mileage has something to do with the price. We are going to use the same loop as previously for top 10 brands, and make a dictionary with the mean value of mileages.

In [154]:
top_10_mileage_mean = {}

for b in top_10_brands:
    top_10_mileage_mean[b] = autos.loc[autos["brand"] == b, "odometer_km"].mean()
    
top_10_mileage_mean

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056,
 'renault': 128071.33121308497,
 'peugeot': 127153.62526920316,
 'fiat': 117121.9715956558,
 'seat': 121131.30128956624}

From the dictionary above we can see that mileage is not a huge factor for prices, as the mileage is fairly even for each of the brands. Let's make a dataframe for easier comparison.

In [172]:
price_series = pd.Series(top_10_price_mean)
mileage_series = pd.Series(top_10_mileage_mean)

df = pd.DataFrame(price_series, columns=["mean_price"])
df["mean_mileage"] = mileage_series
df

mean_price   mean_mileage
volkswagen     5402.410262  128707.158791
bmw            8332.820518  132572.513140
opel           2975.241935  129310.035842
mercedes_benz  8628.450366  130788.363313
audi           9336.687454  129157.386785
ford           3749.469507  124266.012872
renault        2474.864607  128071.331213
peugeot        3094.017229  127153.625269
fiat           2813.748538  117121.971596
seat           4397.230950  121131.301290

The mean mileage for top 10 brands is fairly even, and does not have noticable impact on the mean price.